docs/propagation.md

```diff
@@ -0,0 +1,201 @@
+
+1. Project DimSharding to FactorSharding
+2. Propagate sharding axes in the space of FactorSharding
+3. Projecte the updated FactorSharding to get the updated DimSharding
+
+#### Visualization of Sharding Propagation Along Factors
+
+We will use the following table to visualize the sharding propagation problem and algorithm.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 |  |  |  |  |
+| T1 |  |  |  |  |
+| T2 |  |  |  |  |
+
+* Each column represents a factor. F0 means the factor with index 0. We propagate shardings along factors (columns).
+* Each row represents a tensor. T0 refers to the tensor with index 0. Tensors are all operands and results involved for a specific operation. The axes in a row cannot overlap. An axis (or sub-axis) cannot be used to partition one tensor many times. If an axis is explicitly replicated, we cannot use it to partition the tensor.
+
+Thus, each cell represents a factor sharding. A factor can be missing in partial tensors. The table for `C = dot(A, B)` is below. The highlighted cells imply that the factor is not in the tensor. For example, F2 is in T1 and T2, but not in T0.
+
+| `C = dot(A, B)` | F0 Batching dim | F1 Non-contracting dim | F2 Non-contracting dim | F3 Contracting dim | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- | :---- |
+| T0 = A |  |  |  |  |  |
+| T1 = B |  |  |  |  |  |
+| T2 = C |  |  |  |  |  |
+
+#### Collect and propagate sharding axes
+
+We use a simple example shown below to visualize the propagation.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 | “a” |  | “f” |  |
+| T1 | “a”, “b” | “c”, “d” | “g” |  |
+| T2 |  | “c”, “e” |  |  |
+
+**Step 1\.** Find axes to propagate along each factor (a.k.a. the (longest) compatible major sharding axes). For this example, we propagate \[“a”, “b”\] along F0, propagate \[“c”\] along F1, and propagate nothing along F2.
+
+**Step 2\.** Expand the factor shardings to obtain the following result.
+
+|  | F0 | F1 | F2 | Explicitly replicated axes |
+| :---- | :---- | :---- | :---- | :---- |
+| T0 | “a”, **“b”** | **“c”** | “f” |  |
+| T1 | “a”, “b” | “c”, “d” | “g” |  |
+| T2 | **“a”, “b”** | “c”, “e” |  |  |
+
+### Operations that are treated differently during propagation (**outdated**)
+
+The above propagation step description applies to every op other than `CustomCallOp`, `OptimizationBarrierOp`, `WhileOp`, and `CaseOp`. Here we will talk about how and why they are treated differently.
+
+#### OptimizationBarrierOp
+
+[See StableHLO for the spec of the op](https://github.com/openxla/stablehlo/blob/main/docs/spec.md#optimization_barrier). This op can be thought of as being 1:1 identity mappings between the operand and result, with there being no relationship between how any two operand-result pairs `i` and `j` are sharded.
+
+##### Why an Operation sharding rule doesn’t work
```

Comment:
This section actually is still relevant as it explains why we can treat those ops like others.